trieste/models/keras/architectures.py

```diff
@@ -0,0 +1,315 @@
+            :class:`~tf.keras.layers.Dense` hidden layer. Please check Keras Dense layer API for
+            available arguments. Objects in the list will sequentially be used to add
+            :class:`~tf.keras.layers.Dense` layers. Length of this list determines the number of
+            hidden layers in the network. Default value is two hidden layers, 50 nodes each, with
+            ReLu activation functions. Empty list needs to be passed to have no hidden layers.
+        :param independent: If set to `True` then :class:`~tfp.layers.IndependentNormal` layer
+            is used as the output layer. This models outputs as independent, only the diagonal
+            elements of the covariance matrix are parametrized. If left as the default `False`,
+            then :class:`~tfp.layers.MultivariateNormalTriL` layer is used where correlations
+            between outputs are learned as well.
+        :raise ValueError: If objects in ``hidden_layer_args`` are not dictionaries.
+        """
+        if hidden_layer_args is None:
+            hidden_layer_args = [
+                {"units": 50, "activation": "relu"},
+                {"units": 50, "activation": "relu"},
+            ]
+        if not isinstance(hidden_layer_args, list):
+            raise ValueError(
+                f"hidden_layer_args must be a list of dictionaries."
+                f"Received {type(hidden_layer_args)} instead."
+            )
+        else:
+            if len(hidden_layer_args) > 0:
+                for layer in hidden_layer_args:
+                    if not isinstance(layer, dict):
+                        raise ValueError(
+                            f"Objects in hidden_layer_args must be dictionaries. They should "
+                            f"contain arguments for hidden dense layers, "
+                            f"received {type(layer)} instead."
+                        )
+
+        super().__init__(input_tensor_spec, output_tensor_spec)
+
+        self._num_hidden_layers = len(hidden_layer_args)
+        self._hidden_layer_args = hidden_layer_args
+        self._independent = independent
+
+    def _gen_input_tensor(self) -> tf.keras.Input:
+
+        input_tensor = tf.keras.Input(
+            shape=self.input_tensor_spec.shape,
+            dtype=self.input_tensor_spec.dtype,
+            name=self.input_layer_name,
+        )
+        return input_tensor
+
+    def _gen_hidden_layers(self, input_tensor: tf.Tensor) -> tf.Tensor:
+
+        if self._num_hidden_layers > 0:
+            for index, hidden_layer_args in enumerate(self._hidden_layer_args):
+                layer_name = self.network_name + "dense_" + str(index)
+                layer = tf.keras.layers.Dense(**hidden_layer_args, name=layer_name)
+                input_tensor = layer(input_tensor)
+
+        return input_tensor
+
+    def _gen_output_layer(self, input_tensor: tf.Tensor) -> tf.Tensor:
+
+        if self._independent:
```

Comment:
You could significantly simplify this function by starting with

```python
    layer = tfp.layers.IndependentNormal if self._independent else tfp.layers.MultivariateNormalTriL
```

and using that from then on.

Reply:
good tip!